Error: index out of bounds

```
2019-04-08 13:10:31,065 [Analytics] [ERROR]  handle_analytic_task Exception: 'Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/pandas/core/indexes/base.py", line 2477, in get_value    
    tz=getattr(series.dtype, 'tz', None))
  File "pandas/_libs/index.pyx", line 98, in pandas._libs.index.IndexEngine.get_value (pandas/_libs/index.c:4404)
  File "pandas/_libs/index.pyx", line 106, in pandas._libs.index.IndexEngine.get_value (pandas/_libs/index.c:4087)
  File "pandas/_libs/index.pyx", line 154, in pandas._libs.index.IndexEngine.get_loc (pandas/_libs/index.c:5126)
  File "pandas/_libs/hashtable_class_helper.pxi", line 1210, in pandas._libs.hashtable.PyObjectHashTable.get_item (p
andas/_libs/hashtable.c:20523)
  File "pandas/_libs/hashtable_class_helper.pxi", line 1218, in pandas._libs.hashtable.PyObjectHashTable.get_item (p
andas/_libs/hashtable.c:20477)
KeyError: 0

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "bin/../analytics/analytic_unit_manager.py", line 82, in handle_analytic_task
    result_payload = await self.__handle_analytic_task(task)
  File "bin/../analytics/analytic_unit_manager.py", line 75, in __handle_analytic_task
    return await worker.do_detect(data, payload['cache'])
  File "bin/../analytics/analytic_unit_worker.py", line 67, in do_detect
    detected = self._detector.detect(chunk_dataframe, cache)
  File "bin/../analytics/detectors/pattern_detector.py", line 60, in detect
    detected = self.model.detect(dataframe, self.analytic_unit_id, cache)
  File "bin/../analytics/models/model.py", line 105, in detect
    result = self.do_detect(dataframe, id)
  File "bin/../analytics/models/peak_model.py", line 76, in do_detect
    for i in utils.exponential_smoothing(data + self.state['confidence'], EXP_SMOOTHING_FACTOR):
  File "bin/../analytics/utils/common.py", line 20, in exponential_smoothing
    result = [series[0]]
  File "/usr/local/lib/python3.6/site-packages/pandas/core/series.py", line 601, in __getitem__
    result = self.index.get_value(self, key)
  File "/usr/local/lib/python3.6/site-packages/pandas/core/indexes/base.py", line 2483, in get_value
    return libts.get_value_box(s, key)
  File "pandas/_libs/tslib.pyx", line 923, in pandas._libs.tslib.get_value_box (pandas/_libs/tslib.c:18843)
  File "pandas/_libs/tslib.pyx", line 939, in pandas._libs.tslib.get_value_box (pandas/_libs/tslib.c:18560)
IndexError: index out of bounds
'
```

--- a/analytics/analytics/models/model.py
+++ b/analytics/analytics/models/model.py
@@ -76,12 +76,16 @@ class Model(ABC):
             if segment_map['labeled'] or segment_map['deleted']:
                 segment = Segment(dataframe, segment_map, self.find_segment_center)
                 if segment.percent_of_nans > 0.1 or len(segment.data) == 0:
+                    logging.debug(f'segment {segment_map.start}-{segment_map.end} skip because of invalid data')
                     continue
                 if segment.percent_of_nans > 0:
                     segment.convert_nan_to_zero()
                 max_length = max(segment.length, max_length)
                 if segment.labeled: labeled.append(segment)
                 if segment.deleted: deleted.append(segment)
+
+        assert len(labeled) > 0, f'labeled list empty, skip fitting for {id}'
+
         if self.state.get('WINDOW_SIZE') == 0:            
             self.state['WINDOW_SIZE'] = math.ceil(max_length / 2) if max_length else 0
         model, model_type = self.get_model_type()

--- a/analytics/analytics/utils/dataframe.py
+++ b/analytics/analytics/utils/dataframe.py
@@ -22,6 +22,8 @@ def get_intersected_chunks(data: list, intersection: int, chunk_size: int) -> Ge
         intersection - length of intersection.
         chunk_size - length of chunk
         """
+        assert chunk_size > 0, 'chunk size must be great than zero'
+        assert intersection > 0, 'intersection length must be great than zero'
 
         data_len = len(data)
 
@@ -48,6 +50,7 @@ def get_chunks(data: list, chunk_size: int) -> Generator[list, None, None]:
     Returns generator that splits dataframe on non-intersected segments.
     chunk_size - length of chunk
     """
+    assert chunk_size > 0, 'chunk size must be great than zero'
 
     chunks_iterables = [iter(data)] * chunk_size
     result_chunks = zip(*chunks_iterables)

--- a/analytics/tests/test_dataset.py
+++ b/analytics/tests/test_dataset.py
@@ -20,12 +20,12 @@ class TestDataset(unittest.TestCase):
             models.PeakModel(),
             models.TroughModel()
         ]
-        try:
-            for model in model_instances:
-                model_name = model.__class__.__name__
+
+        for model in model_instances:
+            model_name = model.__class__.__name__
+
+            with self.assertRaises(AssertionError):
                 model.fit(dataframe, segments, 'test', dict())
-        except ValueError:
-            self.fail('Model {} raised unexpectedly'.format(model_name))
     
     def test_peak_antisegments(self):
         data_val = [1.0, 1.0, 1.0, 2.0, 3.0, 2.0, 1.0, 1.0, 1.0, 1.0, 5.0, 7.0, 5.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]

